File: plot/src/main/java/net/fwitz/math/plot/binary/escape/EscapeTimePanel.java
package net.fwitz.math.plot.binary.escape;

import net.fwitz.math.complex.BinaryNumber;
import net.fwitz.math.fractal.escape.EscapeTimeResult;
import net.fwitz.math.plot.binary.BinaryNumberFunctionPanel;
import net.fwitz.math.plot.binary.BinaryNumberFunctionRenderer;

public class EscapeTimePanel<T extends BinaryNumber<T>> extends BinaryNumberFunctionPanel<T , EscapeTimeResult> {
    EscapeTimePanel(Class<T> binaryNumberType, EscapeTimePlot<T> plot) {
        super(binaryNumberType, EscapeTimeResult.class, plot);
    }

    @Override
    protected BinaryNumberFunctionRenderer<T, EscapeTimeResult> createRenderer(int width, int height) {
        return new Renderer(this, width, height);
    }

    class Renderer extends BinaryNumberFunctionRenderer<T, EscapeTimeResult> {
        Renderer(BinaryNumberFunctionPanel<T, EscapeTimeResult> panel, int width, int height) {
            super(panel, EscapeTimeResult.class, width, height, EscapeTimeValuesFilters.getInstance());
        }

        @Override
        protected T value(double x, double y) {
            return plot.value(x, y);
        }
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/binary/escape/EscapeTimePanel.kt
package net.fwitz.math.plot.binary.escape

import net.fwitz.math.binary.BinaryNumber
import net.fwitz.math.fractal.escape.EscapeTimeResult
import net.fwitz.math.plot.binary.BinaryNumberFunctionPanel
import net.fwitz.math.plot.binary.BinaryNumberFunctionRenderer

@Suppress("UNCHECKED_CAST")
class EscapeTimePanel<T : BinaryNumber<T>> internal constructor(
    binaryNumberType: Class<T>,
    plot: EscapeTimePlot<T>
) : BinaryNumberFunctionPanel<T, EscapeTimeResult<T>>(
    binaryNumberType,
    EscapeTimeResult::class.java as Class<EscapeTimeResult<T>>,
    plot
) {
    override fun createRenderer(
        width: Int,
        height: Int
    ): BinaryNumberFunctionRenderer<T, EscapeTimeResult<T>> = Renderer(this, width, height)

    internal inner class Renderer(
        panel: BinaryNumberFunctionPanel<T, EscapeTimeResult<T>>,
        width: Int,
        height: Int
    ) : BinaryNumberFunctionRenderer<T, EscapeTimeResult<T>>(
        panel,
        EscapeTimeResult::class.java as Class<EscapeTimeResult<T>>,
        width,
        height,
        EscapeTimeValuesFilters()
    ) {
        override fun value(x: Double, y: Double) = plot.value(x, y)
    }
}